File: 0146-lru-cache/0146-lru-cache.py
class ListNode:
    def __init__(self, key=0, val=0, next=None, prev=None):
        self.key=key
        self.val=val
        self.next=next
        self.prev=prev
class LRUCache:

    def __init__(self, capacity: int):
        self.cache={}
        self.capacity = capacity
        self.head = ListNode()
        self.tail = ListNode()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: ListNode):
        prev_node = node.prev
        next_node = node.next
        prev_node.next = next_node
        next_node.prev = prev_node
    
    def _remove_least(self, node: ListNode):
        node.prev.next = self.tail
        self.tail.prev = node.prev

    def _add_to_front(self, node: ListNode):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        
        node = self.cache[key]
        self._remove(node)
        self._add_to_front(node)

        return node.val

    def put(self, key: int, value: int) -> None:

        if key not in self.cache:
            new_node = ListNode(key,value)
            self.cache[key] = new_node
            self._add_to_front(new_node)

            if len(self.cache) > self.capacity:
                lru_node = self.tail.prev
                self._remove(lru_node)
                del self.cache[lru_node.key] 
        else:
            node = self.cache[key]
            node.val = value
            self._remove(node)
            self._add_to_front(node)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
private:
    int capacity;

    struct Node{
        int key,value;
        Node* prev;
        Node* next;
        Node(int k, int v): key(k),value(v),prev(nullptr),next(nullptr){}
    };

    unordered_map<int,Node*> cache;
    Node* head;
    Node* tail;

    void removeNode(Node* node){
        node->prev->next=node->next;
        node->next->prev=node->prev;
    }

    void addToHead(Node* node){
        node->prev=head;
        node->next=head->next;
        head->next->prev=node;
        head->next=node;
    }

    void moveToHead(Node* node){
        removeNode(node);
        addToHead(node);
    }

    Node* removeTail(){
        Node* node = tail->prev;
        removeNode(node);
        return node;
    }
public:
    LRUCache(int capacity) {
        this->capacity=capacity;
        head = new Node(0,0);
        tail = new Node(0,0);

        head->next=tail;
        tail->prev=head;
    }
    
    int get(int key) {
        if(cache.count(key)){
            Node* node=cache[key];
            moveToHead(node);
            return node->value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(cache.count(key)){
            Node* node=cache[key];
            node->value=value;
            moveToHead(node);
        }else{
            if(cache.size()==capacity){
                Node* lru=removeTail();
                cache.erase(lru->key);
                delete lru;
            }
            Node* node=new Node(key,value);
            addToHead(node);
            cache[key]=node;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */